### PR TITLE
Fix Issue 21876 - Zero-length static arrays "cannot be read at compile time"

### PR DESCRIPTION
**Pull request by BorisCarvajal**

### CODE REVIEW

File: src/dmd/dinterpret.d [hunk 1]
```diff
@@ -2102,17 +2102,22 @@ public:
...
-                    error(loc, "static variable `%s` cannot be read at compile time", v.toChars());
-                    return CTFEExp.cantexp;
-                }
                 if (!e)
                 {
-                    assert(!(v._init && v._init.isVoidInitializer()));
-                    // CTFE initiated from inside a function
-                    error(loc, "variable `%s` cannot be read at compile time", v.toChars());
-                    return CTFEExp.cantexp;
+                    if (v.type.size() == 0)
```
Comment (UplinkCoder) [inline src/dmd/dinterpret.d:2107]: I am not at all sure this is a good idea.
Also you have removed an assertion.

**Review by UplinkCoder — CHANGES_REQUESTED** [review]
I would like to see a good reason for why it should be allowed, it use erroneous things like int[0] at CTFE.

File: src/dmd/dinterpret.d (hunk 1 above)
Comment (BorisCarvajal) [reply src/dmd/dinterpret.d:2107]: This special case is already here, just look a little above in the [code](https://github.com/dlang/dmd/blob/3db7045051ea8ea7b110ae3bcfbcc005fb2dca05/src/dmd/dinterpret.d#L2082).
Furthermore, the same logic is also used to handle zero-field struct literals, see `visit(StructLiteralExp)` and `visit(DotVarExp)`.

**Comment by BorisCarvajal** [comment]
> I would like to see a good reason for why it should be allowed, it use erroneous things like int[0] at CTFE.

Please add some argument to that.
Actually, the question is why it shouldn't be allowed like in normal code.

**Comment by UplinkCoder** [comment]
This is allowed in normal code?
If so I retract my disapproval.

Can you add a comment above `if (v.type.size() == 0)` 
because it's not oblivious this is checking for `int[0]`

File: src/dmd/dinterpret.d (hunk 1 above)
Comment (UplinkCoder) [reply src/dmd/dinterpret.d:2107]: oh my ...

**Review by UplinkCoder — APPROVED** [review]
This isn't great but since it's already in the code for structs, I guess we can go with this...

**Comment by BorisCarvajal** [comment]
Comment added.

**Review by Geod24 — APPROVED** [review]
Those special cases make me want to bash my head into a wall. Thanks for the added comments!

**Pull request merged**